pennylane/transforms/decorators.py

```diff
@@ -0,0 +1,388 @@
+                fn(*args, **kwargs)
+        return tape
+
+    return wrapper
+
+
+class NonQueuingTape(qml.queuing.AnnotatedQueue):
+    """Mixin class that creates a tape that does not queue
+    itself to the current queuing context."""
+
+    def _process_queue(self):
+        super()._process_queue()
+
+        for obj, info in self._queue.items():
+            qml.queuing.QueuingContext.append(obj, **info)
+
+        qml.queuing.QueuingContext.remove(self)
+
+
+class single_tape_transform:
+    """For registering a tape transform that takes a tape and outputs a single new tape.
+
+    Examples of such transforms include circuit compilation.
+
+    Args:
+        transform_fn (function): The function to register as the single tape transform.
+            It can have an arbitrary number of arguments, but the first argument
+            **must** be the input tape.
+
+    **Example**
+
+    A valid single tape transform is a quantum function that satisfies the following:
+
+    - The first argument must be an input tape
+
+    - Depending on the structure of this input tape, various quantum operations, functions,
+      and templates may be called.
+
+    - Any internal classical processing should use the ``qml.math`` module to ensure
+      the transform is differentiable.
+
+    - There is no return statement.
+
+    For example:
+
+    .. code-block:: python
+
+        @qml.single_tape_transform
+        def my_transform(tape, x, y):
+            # loop through all operations on the input tape
+            for op in tape.operations + tape.measurements:
+                if op.name == "CRX":
+                    wires = op.wires
+                    param = op.parameters[0]
+
+                    qml.RX(x * qml.math.abs(param), wires=wires[1])
+                    qml.RY(y * qml.math.abs(param), wires=wires[1])
+                    qml.CZ(wires=[wires[1], wires[0]])
+                else:
+                    op.queue()
+
+    This transform iterates through the input tape, and replaces any :class:`~.CRX` operation with
+    two single qubit rotations and a :class:`~.CZ` operation. These newly queued operations will
+    form the output transformed tape.
+
+    We can apply this transform to a quantum tape:
+
+    >>> with qml.tape.JacobianTape() as tape:
+    ...     qml.Hadamard(wires=0)
+    ...     qml.CRX(-0.5, wires=[0, 1])
+    >>> new_tape = my_transform(tape, 1., 2.)
+    >>> print(new_tape.draw())
+     0: ──H───────────────╭Z──┤
+     1: ──RX(0.5)──RY(1)──╰C──┤
+    """
+
+    def __init__(self, transform_fn):
+        self.transform_fn = transform_fn
+        functools.update_wrapper(self, transform_fn)
+
+    def __call__(self, tape, *args, **kwargs):
+        tape_class = type(tape.__class__.__name__, (NonQueuingTape, tape.__class__), {})
+
+        with tape_class() as new_tape:
+            self.transform_fn(tape, *args, **kwargs)
+
+        new_tape.__class__ = tape.__class__

```

Comment:
Why is this part necessary if we've already constructed `new_tape` to be of type `tape_class`? (Is it in case the `transform_fn` changes it?)

Reply:
`new_tape`, when first created, is of the class `(NonQueuingTape, tape.__class__)`, so that it doesn't result in a nested tape on the tape. (This solves your issue from the how-to @glassnotes, where composing transforms was leading to nesting).

Just for safety, once we're done I'm reverting it back to be simply an instance of `tape.__class__`.